### PR TITLE
Add an assert concerning branch to CALLFINALLY blocks

### PR DESCRIPTION
**Pull request by BruceForstall**
In the FEATURE_EH_CALLFINALLY_THUNKS case, BBJ_CALLFINALLY blocks live
in the EH region enclosing the `try` block that needs to call the finally.
However, we need all flow to the CALLFINALLY to come from that try block;
we don't want flow optimizations to otherwise branch directly to this
CALLFINALLY block. Add an assert to verify this is the case.

### CODE REVIEW

**Comment by BruceForstall** [comment]
This assert doesn't fire in SPMI, but does fire for the test case here: https://github.com/dotnet/runtime/issues/55131#issuecomment-873570166.

@AndyAyersMS @kunalspathak PTAL
cc @dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
> we don't want flow optimizations to otherwise branch directly to this CALLFINALLY block

I'm not opposed to this, but we should acknowledge this is a choice, not a hard requirement ... unless you know of an actual unfixable correctness issue here.

**Comment by kunalspathak** [comment]
Looks like the assertion is hit during crossgen2 compilation of SPC.dll on windows x86.

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 1]
```diff
@@ -2740,6 +2740,55 @@ void Compiler::fgDebugCheckBBlist(bool checkBBNum /* = false */, bool checkBBRef
...
+                //        try {
+                //            LEAVE L_OUTER; // this becomes a branch to a callfinally in an outer try region (in the
+                //                           // FEATURE_EH_CALLFINALLY_THUNKS case)
+                //        } catch {
+                //        }
+                //    } finally {
+                //    }
+                //    L_OUTER:
+                //
+                assert(bbInTryRegions(finallyIndex, block));
```
Comment (kunalspathak) [inline src/coreclr/jit/fgdiagnostic.cpp:2788]: From your comments, should this be under `FEATURE_EH_CALLFINALLY_THUNKS` ?

**Comment by BruceForstall** [comment]
> I'm not opposed to this, but we should acknowledge this is a choice, not a hard requirement ... unless you know of an actual unfixable correctness issue here.

True, this is a conservative statement about our existing EH flow model. Changing this to be more liberal requires more thinking about the implications. I can add a comment to that effect.

> From your comments, should this be under #ifdef FEATURE_EH_CALLFINALLY_THUNKS ?

It should work for all cases, as the callee does the right thing in all cases.

**Comment by BruceForstall** [comment]
I verified that the new test in https://github.com/dotnet/runtime/pull/55674 hits this assert, and succeeds without hitting this assert with the code fix in that PR is included in the JIT.

**Review by AndyAyersMS — APPROVED** [review]
LGTM.

**Pull request merged**